service/controller/resource/clusterupgrade/create.go

```diff
@@ -134,9 +135,13 @@ func (r *Resource) ensureAzureClusterHasSameRelease(ctx context.Context, cr capi
 
 func (r *Resource) ensureMasterHasUpgraded(ctx context.Context, cluster capiv1alpha3.Cluster) (bool, error) {
 	tenantClusterK8sClient, err := r.tenantClientFactory.GetClient(ctx, &cluster)
-	if err != nil {
-		r.logger.LogCtx(ctx, "level", "debug", "message", "tenant API not available yet", "stack", microerror.JSON(err))
+	if tenant.IsAPINotAvailable(err) {
```

Comment:
I was thinking if it would be better (or worth it) to move the dependency on `tenant` to the local `tenantcluster` package (the one defining the client factory). That way the tenant client factory would use `tenant.IsAPINotAvailable(err)` trying to match the error message, and return its own error object, that handlers would try to match. That way handlers would depend only in the local `tenantcluster` package instead of both `tenantcluster` package and `tenant` library.

Then again, not sure if worth it.

Reply:
It would be worth it. One of the refactoring things I proposed was the caching of credentials and tenant cluster client goes to same direction. Instead of finding TC k8s certs & keys several times within reconciliation, it would be nice to re-use the client already acquired.

Reply:
I gave it a shot, PTAL.